
# Analysis of Doctor Appointments

## Table of Contents
<ol>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ol>



### Question(s) for Analysis
>**Tip**: Clearly state one or more questions that you plan on exploring over the course of the report. You will address these questions in the **data analysis** and **conclusion** sections. Try to build your report around the analysis of at least one dependent variable and three independent variables. If you're not sure what questions to ask, then make sure you familiarize yourself with the dataset, its variables and the dataset context for ideas of what to explore.

> **Tip**: Once you start coding, use NumPy arrays, Pandas Series, and DataFrames where appropriate rather than Python lists and dictionaries. Also, **use good coding practices**, such as, define and use functions to avoid repetitive code. Use appropriate comments within the code cells, explanation in the mark-down cells, and meaningful variable names. 

<a id='intro'></a>
## Introduction

A person makes a doctor appointment, receives all the instructions and no-show. Who to blame? **Well Let's find Out**

### Dataset Description 
This dataset collects information from 100k medical appointments in Brazil and is focused on the question of whether or not patients show up for their appointment. A number of characteristics about the patient are included in each row.
#### Data Dictionary
<dl>
<dt>PatientId</dt>
<dd>Identification of the patient</dd>
<dt>AppointmentID</dt>
<dd>Identification of each appointment</dd>
<dt>Gender</dt>
<dd>Male of Female</dd>

<dt>Age</dt>
<dd>The Age of the patient</dd>

<dt>Naighborhood</dt>
<dd>Where the appointment takes place</dd>

<dt>Scholarship</dt>
<dd>Indicates whether or not the patient is enrolled in the Brasilian welfare program</dd>

<dt>Schedule Day</dt>
<dd>When the appointment was scheduled</dd>

<dt>Hipertension</dt>
<dd>True or false :if patient is suffering from hypertension</dd>

<dt>Diabetes</dt>
<dd>True or False :if patient is suffering from diabetes</dd>

<dt>Alcoholisim</dt>
<dd>True or False :if patient is and alcoholic or not</dd>

<dt>Handcap</dt>
<dd>True of False : if patient is handicapped or not</dd>

<dt>SMS_received</dt>
<dd>Number of messages sent to patient</dd>

<dt>No_show</dt>
<dd>True or False: if patient showed up to the appointment</dd>

</dl>

### Questions for Analysis

1. Is Alcholism likely to affect if a patient will show up for an appointment?
2. What factors are important for us to know in order to predict if a patient will show up for their schedued appointment?

In [1]:
# importing the modules for use
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb


<a id='wrangling'></a>
## Data Wrangling

### General Properties

We will begin our dataset exploration by importing our data and previewing the top of the dataset.

In [2]:
df = pd.read_csv("noshowappointments-kagglev2-may-2016.csv")
df.head(10)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   
5  9.598513e+13        5626772      F  2016-04-27T08:36:51Z   
6  7.336882e+14        5630279      F  2016-04-27T15:05:12Z   
7  3.449833e+12        5630575      F  2016-04-27T15:39:58Z   
8  5.639473e+13        5638447      F  2016-04-29T08:02:16Z   
9  7.812456e+13        5629123      F  2016-04-27T12:48:25Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   
5  2016-04-29T00:00:00Z   76          REPÚBLICA            0             1   
6  2016-04-29T00:00:00Z   23         GOIABEIRAS            0             0   
7  2016-04-29T00:00:00Z   39         GOIABEIRAS            0             0   
8  2016-04-29T00:00:00Z   21         ANDORINHAS            0             0   
9  2016-04-29T00:00:00Z   19          CONQUISTA            0             0   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No  
5         0           0        0             0      No  
6         0           0        0             0     Yes  
7         0           0        0             0     Yes  
8         0           0        0             0      No  
9         0           0        0             0      No

We will then check the number of columns and number of rows that are present in our data.

In [3]:
df.shape

(110527, 14)

We see that our data has 14 columns and 110,527 rows of data. We will then check on the names of our columns.

In [4]:
df.columns

Index(['PatientId', 'AppointmentID', 'Gender', 'ScheduledDay',
       'AppointmentDay', 'Age', 'Neighbourhood', 'Scholarship', 'Hipertension',
       'Diabetes', 'Alcoholism', 'Handcap', 'SMS_received', 'No-show'],
      dtype='object')

Having generated the dataset columns we will then check on the data types of the columns.

In [5]:
df.dtypes

PatientId         float64
AppointmentID       int64
Gender             object
ScheduledDay       object
AppointmentDay     object
Age                 int64
Neighbourhood      object
Scholarship         int64
Hipertension        int64
Diabetes            int64
Alcoholism          int64
Handcap             int64
SMS_received        int64
No-show            object
dtype: object

Our dataset has columns of different datatyeps namely float, integers, and string (we will assign the appropriate data types of these columns later) we will then generate the general information of the dataset.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


### Data Cleaning

Our first step towards effective data cleaning is ensuring that we have no missing values in our data and if there is any ensuring that they are dealed with appropriately.

We will check on missing values in the data.

In [8]:
df.isna().any() == True

PatientId         False
AppointmentID     False
Gender            False
ScheduledDay      False
AppointmentDay    False
Age               False
Neighbourhood     False
Scholarship       False
Hipertension      False
Diabetes          False
Alcoholism        False
Handcap           False
SMS_received      False
No-show           False
dtype: bool

None of teh columns has any missing, hence we will move on to the next step in data cleaning that is ensuring every column has the appropriate data type. We will first begin by checking on the data types of the column.

In [9]:
""" checking on the data types of the columns """
df.dtypes

PatientId         float64
AppointmentID       int64
Gender             object
ScheduledDay       object
AppointmentDay     object
Age                 int64
Neighbourhood      object
Scholarship         int64
Hipertension        int64
Diabetes            int64
Alcoholism          int64
Handcap             int64
SMS_received        int64
No-show            object
dtype: object

From the cell output above we see that the **AppointmentDay** and **ScheduleDay** are of data type object meaning that they are strings, why is this?

Lets visualize the dataset once again?

In [10]:
df.head(10)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   
5  9.598513e+13        5626772      F  2016-04-27T08:36:51Z   
6  7.336882e+14        5630279      F  2016-04-27T15:05:12Z   
7  3.449833e+12        5630575      F  2016-04-27T15:39:58Z   
8  5.639473e+13        5638447      F  2016-04-29T08:02:16Z   
9  7.812456e+13        5629123      F  2016-04-27T12:48:25Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   
5  2016-04-29T00:00:00Z   76          REPÚBLICA            0             1   
6  2016-04-29T00:00:00Z   23         GOIABEIRAS            0             0   
7  2016-04-29T00:00:00Z   39         GOIABEIRAS            0             0   
8  2016-04-29T00:00:00Z   21         ANDORINHAS            0             0   
9  2016-04-29T00:00:00Z   19          CONQUISTA            0             0   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No  
5         0           0        0             0      No  
6         0           0        0             0     Yes  
7         0           0        0             0     Yes  
8         0           0        0             0      No  
9         0           0        0             0      No

From the output above we see that reason is because they contain **T** and **Z** lets clean this up.

>We will begin by cleaning up the Schedule we will split this column on delimeter **T** and renaming the new columns

In [34]:
schedule = df['ScheduledDay'].str.split("T", expand=True)
schedule.columns = ["Schedule_date", "Schedule_Time"]

Having split the data based on the delimeter **T** we will remove the **Z** by splitting it from the rest of the data, we will then add the two columns together to produce a single colum containing the date and time that is clean.

In [ ]:
schedule_day = schedule["Schedule_Time"].str.split("Z", expand=True)
schedule_day.columns = ["Schedule_Time", "drop"]
schd = pd.concat([schedule_day["Schedule_Time"], schedule['Schedule_date']], axis=1)
schd1 = pd.DataFrame(schd['Schedule_date'] + " " +schd['Schedule_Time'])
schd1.columns = ['schedule']

We will then move onto the **AppointmentDate** here we willl split the data based on the **T** and **Z** delimeter.

it will produce two columns one with date and teh other with time, hwoever we will notice that the time column is 0 hence we drop the column to remain with date only.

In [ ]:
#appointment date column
appoint = df['AppointmentDay'].str.split("T", expand=True)
appoint.columns = ["appoint_date", "appoint_Time"]
app = appoint.drop(labels=["appoint_Time"], axis = 1)

We will then convert the concat the schedule and appointment columns and convert them to datetime

In [31]:
schd1['schedule'] = pd.to_datetime(schd1['schedule'])
time_d = pd.concat([schd1, app], axis=1)
time_d['appoint_date'] = pd.to_datetime(time_d["appoint_date"])


It is important to note that the **AppointmetId** and **PatientId** are unique identifiers of the **AppointmentDay** and **ScheduleDay** from our dataset since we will be replacing them with teh cleaned version.

In [32]:
df1 = df.drop(columns = ['AppointmentDay', 'ScheduledDay', 'PatientId', 'AppointmentID'])
df_comp = pd.concat([time_d, df1], axis=1)
#preview the dataset
df_comp.head(10)

schedule appoint_date Gender  Age      Neighbourhood  \
0 2016-04-29 18:38:08   2016-04-29      F   62    JARDIM DA PENHA   
1 2016-04-29 16:08:27   2016-04-29      M   56    JARDIM DA PENHA   
2 2016-04-29 16:19:04   2016-04-29      F   62      MATA DA PRAIA   
3 2016-04-29 17:29:31   2016-04-29      F    8  PONTAL DE CAMBURI   
4 2016-04-29 16:07:23   2016-04-29      F   56    JARDIM DA PENHA   
5 2016-04-27 08:36:51   2016-04-29      F   76          REPÚBLICA   
6 2016-04-27 15:05:12   2016-04-29      F   23         GOIABEIRAS   
7 2016-04-27 15:39:58   2016-04-29      F   39         GOIABEIRAS   
8 2016-04-29 08:02:16   2016-04-29      F   21         ANDORINHAS   
9 2016-04-27 12:48:25   2016-04-29      F   19          CONQUISTA   

   Scholarship  Hipertension  Diabetes  Alcoholism  Handcap  SMS_received  \
0            0             1         0           0        0             0   
1            0             0         0           0        0             0   
2            0             0         0           0        0             0   
3            0             0         0           0        0             0   
4            0             1         1           0        0             0   
5            0             1         0           0        0             0   
6            0             0         0           0        0             0   
7            0             0         0           0        0             0   
8            0             0         0           0        0             0   
9            0             0         0           0        0             0   

  No-show  
0      No  
1      No  
2      No  
3      No  
4      No  
5      No  
6     Yes  
7     Yes  
8      No  
9      No

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. **Compute statistics** and **create visualizations** with the goal of addressing the research questions that you posed in the Introduction section. You should compute the relevant statistics throughout the analysis when an inference is made about the data. Note that at least two or more kinds of plots should be created as part of the exploration, and you must  compare and show trends in the varied visualizations. 



> **Tip**: - Investigate the stated question(s) from multiple angles. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables. You should explore at least three variables in relation to the primary question. This can be an exploratory relationship between three variables of interest, or looking at how two independent variables relate to a single dependent variable of interest. Lastly, you  should perform both single-variable (1d) and multiple-variable (2d) explorations.


### Research Question 1 (Replace this header name!)

In [13]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


### Research Question 2  (Replace this header name!)

In [14]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed in relation to the question(s) provided at the beginning of the analysis. Summarize the results accurately, and point out where additional research can be done or where additional information could be useful.

> **Tip**: Make sure that you are clear with regards to the limitations of your exploration. You should have at least 1 limitation explained clearly. 

> **Tip**: If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work here, check over your report to make sure that it is satisfies all the areas of the rubric (found on the project submission page at the end of the lesson). You should also probably remove all of the "Tips" like this one so that the presentation is as polished as possible.

## Submitting your Project 

> **Tip**: Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).

> **Tip**: Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.

> **Tip**: Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [15]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])

/bin/python: No module named nbconvert


1